### PR TITLE
Fix optimize projections bug

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#8942](https://github.com/apache/arrow-datafusion/issues/8942).

## Rationale for this change
See issue.

## What changes are included in this PR?

Note: This PR is an implementation of @gruuya's suggestion in [discussion](https://github.com/apache/arrow-datafusion/issues/8942#issuecomment-1904175896), with his test in the [PR](https://github.com/apache/arrow-datafusion/pull/8951).

Currently projections are not inserted to logical plan, if its input schema and its schema are same. However, this equality is not a sufficient condition. See #8942 for an example how can this assumption be a problem. On top of that condition, we need to make sure that projection is trivial (e.g it just emits its input). Then we can deem Projection unnecessary. 

However, adding `.iter().all(is_trivial)` check inserts additional `ProjectionExec`s to the plan for 2 test cases as observed by @gruuya. Even if these test plans retract, I think we should do this change. Because removing `Projection` in those plan overfits the name being same after casting (this assumption is not safe). 

I think, what we should really do for those queries is rewriting window expression so that they no longer contain CAST expr after once their input have casted result. I think that is the concern of another PR.

@gruuya suggested [PR](https://github.com/apache/arrow-datafusion/pull/8951) for fixing this issue without retracting these window tests. However, I think that PR solves the problem only for consecutive projections. Hence even if 2 tests regresses we should continue with these changes

## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/sqllogictest/test_files/window.slt [hunk 1]
```diff
@@ -2947,25 +2947,34 @@ logical_plan
...
...data_infinite2.b ASC NULLS LAST, annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.a, annotated_data_infinite2.d] ORDER BY [annotated_data_infinite2.b ASC NULLS LAST, annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 1 FOLLOWING AND 5 FOLLOWING, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.b, annotated_data_infinite2.a] ORDER BY [annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.b, annotated_data_infinite2.a] ORDER BY [annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 5 PRECEDING AND 5 FOLLOWING, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.b, annotated_data_infinite2.a, annotated_data_infinite2.d] ORDER BY [annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN 2 PRECEDING AND 1 FOLLOWING, SUM(annotated_data_infinite2.c) PARTITION BY [annotated_data_infinite2.b, annotated_data_infinite2.a, annotated_data_infinite2.d] ORDER BY [annotated_data_infinite2.c ASC NULLS LAST] ROWS BETWEEN CURRENT ROW AND 1 FOLLOWING
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/window.slt:2950]: In these tests, there is additional Projections with Cast exprs

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo  -- this makes sense to me. Maybe @sergiimk  / @gruuya  have some time to review as well

I think it might be valuable to add the tests case from @sergiimk that verifies the actual values from https://github.com/apache/arrow-datafusion/issues/8942#issue-2092915243 as part of the dataframe tests as well

https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/dataframe/mod.rs

I can do so as well as a follow on PR if you prefer.

**Review by gruuya — APPROVED** [review]
LGTM!

It would be nice to do a follow-up on the extra projections in the window SLTs now. Since the results haven't changed presumably these are redundant, so there should be a heuristic to optimize them away.

**Review by sergiimk — APPROVED** [review]
To my limited experience with DF internals this looks good :+1:

**Comment by mustafasrepo** [comment]
> I think it might be valuable to add the tests case from @sergiimk that verifies the actual values from [#8942 (comment)](https://github.com/apache/arrow-datafusion/issues/8942#issue-2092915243) as part of the dataframe tests as well
> 
> https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/dataframe/mod.rs
> 
> I can do so as well as a follow on PR if you prefer.
 
I have added test in the issue as dataframe test

**Comment by alamb** [comment]
Thanks everyone@

File: datafusion/optimizer/src/optimize_projections.rs [hunk 2]
```diff
@@ -867,7 +867,9 @@ fn rewrite_projection_given_requirements(
     return if let Some(input) =
         optimize_projections(&proj.input, config, &required_indices)?
     {
-        if &projection_schema(&input, &exprs_used)? == input.schema() {
+        if &projection_schema(&input, &exprs_used)? == input.schema()
+            && exprs_used.iter().all(is_expr_trivial)
```
Comment (joroKr21) [inline datafusion/optimizer/src/optimize_projections.rs:871]: A few lines below (882) we have the same check

**Pull request merged**